### PR TITLE
feat: Add Rainbow Brush

### PR DESCRIPTION
**Pull request by vivek-30**
Fixes HarshKhandeparkar/rainbow-board#3
try to add a new brush i.e., Gradient Brush.

### CODE REVIEW

File: src/renderers/RealDrawBoard/RealDrawBoard.ts [hunk 1]
```diff
@@ -39,7 +39,7 @@ export class RealDrawBoard extends RealRenderer {
   eraserSize: number;
   tool: Tool = RealDrawBoardDefaults.tool;
   _isDrawing: boolean = false;
-  _: boolean = false; // If a tool is drawing a strwoke
+  _: boolean = false; // If a tool is drawing a stroke
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/RealDrawBoard.ts:42]: Is this variable even used :joy:?

File: src/renderers/RealDrawBoard/tools/gradientBrush.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import { RealDrawBoard } from '../RealDrawBoard';
+import { Texture } from 'gpu.js';
+import { convertHSLToRGB } from '../../../util/convertForm';
+import { type } from 'os';
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/tools/gradientBrush.ts:4]: I think this import is a mistake.

File: src/renderers/RealDrawBoard/tools/gradientBrush.ts [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+var hue: number = 0;
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/tools/gradientBrush.ts:6]: Let's use `let` to make it modern :smiley:

File: src/renderers/RealDrawBoard/tools/gradientBrush.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+var gradientColors: [number ,number ,number] = [ 1 ,1 ,1 ]; 
+
+export const name = 'gradientBrush';
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/tools/gradientBrush.ts:9]: Can we name this `gradient-brush`?

File: src/util/convertForm.ts [hunk 5]
```diff
@@ -13,11 +13,51 @@ export function convertPolarCartesian(r: number, theta: number) {
 /**
  * Convert Cartesian to polar form.
  * @param x Real Part
- * @param theta Complex Part
+ * @param y Complex Part
```
Comment (HarshKhandeparkar) [inline src/util/convertForm.ts:16]: Thanks, lol.

File: src/util/convertForm.ts [hunk 6]
```diff
@@ -13,11 +13,51 @@ export function convertPolarCartesian(r: number, theta: number) {
...
+}
+
+/**
+ * Convert hsl Color into RGB color.
+ * @param h color value ranges from 0 to 360.
+ * @param s saturation value.
+ * @param l brightness level.
+ * @returns array containing r g and b value
+ */
+export function convertHSLToRGB(h: number, s: number, l: number) :[ number ,number ,number ] {
```
Comment (HarshKhandeparkar) [inline src/util/convertForm.ts:32]: Can this be put in a separate file?

**Review by HarshKhandeparkar — CHANGES_REQUESTED** [review]
Small initial changes.

File: src/renderers/RealDrawBoard/RealDrawBoard.ts (hunk 1 above)
Comment (vivek-30) [reply src/renderers/RealDrawBoard/RealDrawBoard.ts:42]: this is not created by me it is just i have seen a typo that's why i have corrected this

File: src/renderers/RealDrawBoard/tools/gradientBrush.ts (hunk 3 above)
Comment (vivek-30) [reply src/renderers/RealDrawBoard/tools/gradientBrush.ts:6]: ok sure

File: src/renderers/RealDrawBoard/tools/gradientBrush.ts (hunk 4 above)
Comment (vivek-30) [reply src/renderers/RealDrawBoard/tools/gradientBrush.ts:9]: sure

Comment (vivek-30) [reply src/renderers/RealDrawBoard/tools/gradientBrush.ts:9]: sure

File: src/util/convertForm.ts (hunk 6 above)
Comment (vivek-30) [reply src/util/convertForm.ts:32]: ok please suggest a good file name

File: src/renderers/RealDrawBoard/tools/gradientBrush.ts (hunk 2 above)
Comment (vivek-30) [reply src/renderers/RealDrawBoard/tools/gradientBrush.ts:4]: Yes .this is my mistake i will fix it😅

**Comment by vivek-30** [comment]
@HarshKhandeparkar did all the required changes please have a look on them 😊

File: src/renderers/RealDrawBoard/tools/gradient_brush.ts [hunk 7]
```diff
@@ -0,0 +1,53 @@
+import { RealDrawBoard } from '../RealDrawBoard';
+import { Texture } from 'gpu.js';
+import { convertHSLToRGB } from '../../../util/convertHSLToRGB';
+
+let hue: number = 0;
+let gradientColors: [number ,number ,number] = [ 1 ,1 ,1 ];
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/tools/gradient_brush.ts:6]:
```suggestion
let gradientColors: [number, number, number] = [1, 1, 1]; 
```

File: src/renderers/RealDrawBoard/tools/gradient_brush.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+let gradientColors: [number ,number ,number] = [ 1 ,1 ,1 ]; 
+
+export const name = 'gradient_brush';
+
+export function _startStroke(
+  this: RealDrawBoard,
+  coords: [number, number],
+  identifier: string
+) {
+  gradientColors = convertHSLToRGB(hue,90,40);
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/tools/gradient_brush.ts:15]:
```suggestion
  gradientColors = convertHSLToRGB(hue, 90, 40);
```

File: src/renderers/RealDrawBoard/tools/gradient_brush.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+  this._doPreview = false;
+  this._plot(coords[0], coords[1], this.brushSize, gradientColors);
+}
+
+export function _endStroke(
+  this: RealDrawBoard,
+  endCoords: [number, number],
+  identifier: string
+) {
+  gradientColors = convertHSLToRGB(hue,90,40);
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/tools/gradient_brush.ts:25]:
```suggestion
  gradientColors = convertHSLToRGB(hue, 90, 40);
```

File: src/renderers/RealDrawBoard/tools/gradient_brush.ts [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+  this._plot(endCoords[0], endCoords[1], this.brushSize, gradientColors);
+  this._doPreview = true;
+}
+
+export function _doStroke(
+  this: RealDrawBoard,
+  coords: [number, number],
+  identifier: string
+) {
+  hue = (hue+1)%360;
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/tools/gradient_brush.ts:35]:
```suggestion
  hue = (hue + 1) % 360;
```

File: src/renderers/RealDrawBoard/tools/gradient_brush.ts [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+  gradientColors = convertHSLToRGB(hue,90,40);
```
Comment (HarshKhandeparkar) [inline src/renderers/RealDrawBoard/tools/gradient_brush.ts:36]:
```suggestion
  gradientColors = convertHSLToRGB(hue, 90, 40);
```

File: src/util/convertHSLToRGB.ts [hunk 12]
```diff
@@ -0,0 +1,39 @@
+/**
+ * Convert hsl Color into RGB color.
+ * @param h color value ranges from 0 to 360.
+ * @param s saturation value.
+ * @param l brightness level.
+ * @returns array containing r g and b value
+ */
+export function convertHSLToRGB(h: number, s: number, l: number) :[ number ,number ,number ] {
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:8]:
```suggestion
export function convertHSLToRGB(h: number, s: number, l: number): [number, number, number] {
```

File: src/util/convertHSLToRGB.ts [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+    let c: number = (1 - Math.abs(2 * l - 1)) * s;
+    let x: number = c * (1 - Math.abs((h / 60) % 2 - 1));
+    let m: number = l - c/2;
+    let r: number = 0;
+    let g: number = 0;
+    let b: number = 0;
+
+    if (0 <= h && h < 60) {
+        r = c; g = x; b = 0;  
+    } else if (60 <= h && h < 120) {
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:22]:
```suggestion
    }
    else if (60 <= h && h < 120) {
```

File: src/util/convertHSLToRGB.ts [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+        r = x; g = c; b = 0;
+    } else if (120 <= h && h < 180) {
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:24]:
```suggestion
    }
    else if (120 <= h && h < 180) {
```

File: src/util/convertHSLToRGB.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+        r = 0; g = c; b = x;
+    } else if (180 <= h && h < 240) {
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:26]:
```suggestion
    }
    else if (180 <= h && h < 240) {
```

File: src/util/convertHSLToRGB.ts [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+        r = 0; g = x; b = c;
+    } else if (240 <= h && h < 300) {
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:28]:
```suggestion
    }
    else if (240 <= h && h < 300) {
```

File: src/util/convertHSLToRGB.ts [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+        r = x; g = 0; b = c;
+    } else if (300 <= h && h < 360) {
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:30]:
```suggestion
    }
    else if (300 <= h && h < 360) {
```

File: src/util/convertHSLToRGB.ts [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+        r = c; g = 0; b = x;
+    }
+    
+    r = Math.round((r + m));
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:34]:
```suggestion
    r = Math.round(r + m);
```

File: src/util/convertHSLToRGB.ts [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+    g = Math.round((g + m));
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:35]:
```suggestion
    g = Math.round(g + m);
```

File: src/util/convertHSLToRGB.ts [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    b = Math.round((b + m));
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:36]:
```suggestion
    b = Math.round(b + m);
```

File: src/util/convertHSLToRGB.ts [hunk 21]
```diff
@@ -0,0 +1,39 @@
...
+
+    return [ r ,g ,b ];
```
Comment (HarshKhandeparkar) [inline src/util/convertHSLToRGB.ts:38]:
```suggestion
    return [r, g, b];
```

**Review by HarshKhandeparkar — CHANGES_REQUESTED** [review]
Just to maintain consistency :)

**Comment by HarshKhandeparkar** [comment]
The scariest part is still left: Docs!

**Comment by vivek-30** [comment]
> The scariest part is still left: Docs!

Sorry, but i didn't get this 😅 .

**Review by HarshKhandeparkar — CHANGES_REQUESTED** [review]
Only line 248 in the README is left to be changed. After which, we can merge this :smiley:

**Comment by vivek-30** [comment]
@HarshKhandeparkar please have a look now ☺️

**Comment by vivek-30** [comment]
@HarshKhandeparkar please review these changes. hope you likes them 😊

**Review by HarshKhandeparkar — APPROVED** [review]

**Comment by HarshKhandeparkar** [comment]
Thanks!

**Comment by vivek-30** [comment]
Welcome @HarshKhandeparkar and thanks to you as well 😊 without your help it won't be possible

**Pull request merged**